### ISSUE TITLE
We're changing Roc's license, and we need your sign-off to do it!

### ISSUE THREAD

**Issue by rtfeldman** [roc-lang/roc#1199]
**tl;dr** It turns out Roc's Apache2 license means GPLv2 projects won't legally be able to use Roc. To fix this, we're relicensing, which legally requires sign-off from everyone who's contributed code! (Below this is a long explanation about why the change from Apache2 to UPL specifically, if you're interested.) If you're okay giving your permission to relicense your contributions to Roc under the new license, the quickest way to formally do that is to copy/paste this into a comment below this issue:

> I hereby give my permission to license my past and future contributions to this repository under the Universal Permissive License (UPL) Version 1.0 - https://opensource.org/licenses/UPL

We've been talking about it [on Zulip](roc.zulipchat.com/), so feel free to drop by if you have questions. (Also, if it's been awhile, in general feel free to drop by anytime to say hi and check out what's been going on with the project!)

Anyway, here's the long version:

## The Problem

Almost all open-source licenses fall under one of three categories:

1. **Permissive** (e.g. MIT, Apache2) - These are the most common. They typically say anyone can use the software as long as they give attribution.
2. **Copyleft** (e.g. GPL, EUPL) - These are the next-most common. They typically say that anyone can use the software with the caveat that using it (and/or modifying it) may require them to release their code under the same license.
3. **Public Domain** (e.g. Unlicense, CC0) - These are the least common. They typically say anyone can use the software even without attribution.

I wanted Roc to have a permissive license, and I chose Apache2 based on my understanding that it was the best choice because it explicitly [includes a patent protection clause](https://opensource.com/article/18/2/how-make-sense-apache-2-patent-license) that the more concise [MIT and BSD licenses do not](http://web.archive.org/web/20190312190935/https://stlr.org/2018/10/15/the-truth-about-oss-frand-by-all-indications-compatible-models-in-standards-settings/).
This clause theoretically prevents contributors from coming back later and saying "yeah I granted my permission to use this code I wrote from a copyright perspective, but I also have a patent on that code I contributed, and I never said you could use the patent, so I'm going to sue you for patent infringment for using this code."

> **Aside:** Apparently there is debate as to whether [the explicit patent clause is necessary](https://opensource.stackexchange.com/questions/7964/why-is-the-apache-license-2-0-patent-license-clause-useful-important), as well as [whether this clause would actually be an effective lawsuit deterrent in practice](https://www.hodlerlaw.com/2013/12/02/the-apache-license-version-2-0-and-the-anti-patent-treachery-clause/), but there's certainly consensus that Apache2 is strictly safer to use when it comes to patents.

What I didn't realize is that Apache2 has a major downside: although it's compatible with the GPL version 3, it's [incompatible with GPLv2](https://www.apache.org/licenses/GPL-compatibility.html) - the license used by Linux, among many other projects.

This would mean that a lot of GPLv2 projects out there couldn't use Roc's standard library, which in turn means there basically couldn't be any Roc platforms for them. That would be a shame!

## How to fix this?

There are a few ways to fix this:

1. Relicense to MIT (or similar) and give up Apache2's "you won't get sued for patent infringement for using this software" protection.
2. Relicense to "Apache2 or MIT, at your choice", like Rust does; this allows GPLv2 projects to use the MIT license, and everyone else to use the stronger Apache2 license.
3. Relicense to [a modified Apache2, like LLVM did](https://blog.gerv.net/2016/09/gplv2-combination-exception-for-the-apache-2-license/) - which would mean Roc would have to be using an "Apache2 with modifications" license, which I think sounds unacceptably suspicious compared to using a stock license off-the-shelf.
4. Relicense to the [BSD+Patent](https://opensource.org/licenses/BSDplusPatent) license, which fixes this problem but unfortunately [shares a name with an infamous Facebook license](https://www.hackernoon.com/facebooks-bsd-patents-license-and-how-it-affects-you-66088e052845), causing seemingly unavoidable confusion.
5. Relicense to the [Universal Permissive License (UPL)](https://opensource.org/licenses/UPL), which [was created to fix this problem](https://oss.oracle.com/licenses/upl/#faq-1).

Although UPL is not widely used, it seems to have a lot going for it:

* It's concise and permissive, like MIT
* It has patent protections like Apache2
* It's [FSF-approved](https://www.fsf.org/blogs/licensing/universal-permissive-license-added-to-license-list), it's [OSI-approved](https://opensource.org/licenses/UPL), and it has also been vetted by Oracle lawyers, where it was created. That it's been approved by groups with such prominence, and with such conflicting agendas, inspires confidence to me.

Having one license that meets all these criteria seems like a simpler and nicer solution than Rust's dual-licensing, and evidently the people who made that decision in Rust [would have preferred to use a single license if one had existed at the time](https://internals.rust-lang.org/t/rationale-of-apache-dual-licensing/8952/5) which had both patent protections and GPLv2 compatibility.

## Making a Decision

I think the edge case is worth solving; I want to change Roc's license so people can use Roc code with GPLv2.

The most appealing ways to do that are:

1. Relicense to UPL
2. Relicense to MIT
3. Relicense to "either Apache2 or MIT, at your choice"

My impression is that most people either assume that any sufficiently popular open-source project is licensed in a reasonable way, and don't really think about it, or else they check whether it's copyleft or permissive and call it a day. Any stock license (i.e. doesn't say "with modifications") that shows up on GitHub as an [approved first-class license](https://github.com/rtfeldman/dotfiles/blob/trunk/LICENSE) probably will not be given much thought by anyone using Roc. So I don't think choosing any of these options would meaningfully affect Roc's adoption.

Relicensing to MIT would not be the end of the world. Lots of great projects use MIT, and likely almost nobody would complain about it. However, it would always bother me that I could have had explicit patent protections for Roc users and chose not to.

The Apache2/MIT dual license would be strictly safer from a legal perspective. Both licenses are widely known and trusted. It's what Rust does, so there's precedent for it. However, it's also strictly worse from a usability perspective, and it would always bother me that instead of having a single license, we foisted a decision on our users ("Which license do you want to use? Not sure which is right for you? Time to learn about software law!") to accommodate a license (GPLv2) we aren't even using.

The UPL is a single license, so nobody is presented with a choice they feel unequipped to make. It has patent protections. It has GPLv2 compatibility. As a bonus, it's also much more concise than Apache2. That said, UPL does have downsides:

1. It's not as widely used as Apache2 or MIT.
2. It came out of Oracle, and Oracle has a bad reputation when it comes to software law.

I don't think it's a big deal that it isn't widely used. As previously mentioned, I don't think most people pay much attention to software licenses in practice; instead, they (correctly) assume if the project is sufficiently popular, others who care more about licenses will have raised the alarm if the license had serious problems. That's what happened with Facebook's "BSD+Patents" license, after all. Since UPL has basically all the most important seals of approval a software license can ask for (OSI, FSF, and a giant commercial entity), it seems unlikely that it will turn out to have serious problems.

As for the Oracle part, one thought that occurred to me was "Wow, FSF and OSI must have looked *extra closely* at this before giving their approval." If ever there were a license they'd be eager to label deficient, it would be one that came from Oracle! To me, the fact that they both approved it speaks volumes. Especially given the incentives involved, it's hard to argue with the legal evaluations of the two main organizations in the software licensing community.

Overall, I find the unusual choice of going with UPL most appealing here. Among the options, I like its downsides the best.

I ran this idea by a variety of different people, both inside and outside the Roc community, and the consistent response was that relicensing Roc to UPL seemed like the right choice for the project.

## Thank you so much!

I really appreciate your taking the time to do this, on top of the time you've already put into Roc. It means a lot to me!

Here is the full list of contributors being contacted about this - feel free to add the check next your name after copy/pasting the comment below! The comment, for reference:

> I hereby give my permission to license my past and future contributions to this repository under the Universal Permissive License (UPL) Version 1.0 - https://opensource.org/licenses/UPL

- [x] @rtfeldman
- [x] @folkertdev
- [x] @Chadtech
- [x] @Anton-4
- [x] @rvcas
- [x] @bhansconnect
- [x] @sebsheep
- [x] @jaredramirez
- [x] @stoeffel
- [x] @EricHenry
- [x] @celsobonutti
- [x] @pablohirafuji
- [x] @pehota
- [x] @danbruder
- [x] @Arkham
- [x] @pithub
- [x] @pcfields
- [x] @HarrisonB
- [x] @sladwig
- [x] @supermario
- [x] @popara
- [x] @akselw
- [x] @bob-shelline
- [x] @g-belmonte
- [x] @ianmackenzie
- [x] @wpiekutowski
- [x] @wolfadex
- [x] @nnnnathann
